Add support for more precise PreferredWord placement in Rename patterns

Alanoll opened the pull request:
#### Database Migration
NO

#### Description
Episode Naming formats currently support the {Preferred Words} token, which combines all matched preferred words from Release Profiles and replaces the token with them.

This PR adds support to specific Preferred Words to be more precise in the placement of the naming pattern.

Example Pattern:

> {Series Title}.s{season:00}e{episode:00}.{absolute:000}.{Preferred Words EpisodeVersion}.{Episode Title}.{Quality Title}.{Preferred Words}.{Release Group}

Example Release Profile
<img src="https://user-images.githubusercontent.com/21229471/115629133-5a2ca900-a2c7-11eb-9071-828f24228db5.png" width="200">


#### Todos
- [x] Tests
- [ ] Wiki Updates


#### Issues Fixed or Closed by this PR

*

Alanoll commented:
All the tests in the Tests projects passed, but I didn't see any specific tests for existing Preferred Words functionality.

markus101 commented:
This is already part of quality, I think a better solution here would be to have a `Quality Revision` or `Quality Version` which would handle the v# when greater than 1, that way users that don't use preferred words and have propers enabled will benefit as well.

Alanoll commented:
Perhaps Quality was a bad example, it was just the one in my mind at the moment.
This could be used for tagging sources as well like from AMZN/HULU/DSNP

At the moment, any and all Preferred Words can on be placed in location in the name pattern. 
Say you wanted two preferred words:
- SourceService (e.g. AMZN, HULU, etc)
- InternalProperRepack (e.g. You want to respect those three words. The current Quality options convert Proper and Repack both to Proper)

But I'd like the file name to be:
SeriesName.s00e00.EpisodeTitle.{SourceService}-ReleaseGroup-{InternalProperRepack}

I can't do that with the current preferred words. Both sets would be listed in the single location.

Alanoll commented:
> 
> 
> This is already part of quality, I think a better solution here would be to have a `Quality Revision` or `Quality Version` which would handle the v# when greater than 1, that way users that don't use preferred words and have propers enabled will benefit as well.

To further prove it was a bad example on my part, `Quality Proper` already does the version number if the Series Type is Anime.

Taloth commented:
@markus101 I don't like dynamic tags. I suggest using the formatter instead `{Preferred Words:EpisodeTitle}` That way it follows the same approach as season, episode, languages etc.

Alanoll commented:
I took @Taloth 's suggestion and converted to using the formatter pattern instead. I squashed the commits to keep the history clean.

Here's a more concrete example of what this can help users achieve:

Releases:
`The.Office.US.S07E20.EXTENDED.1080p.BluRay.x264-DEiMOS`
`The.Resident.S04E06.Requiems.Revivals.1080p.AMZN.WEB-DL.DDP5.1.H.264-KiNGS`

I can get close to the scene naming patterns, but it won't be perfect. For these examples, focus on the `EXTENDED` and the `AMZN` tags. I want to keep them relative to their original placements (e.g. `EXTENDED` is before quality, and `AMZN` is after quality).

Current Supported Renaming Pattern:
`{Series Title}.s{season:00}e{episode:00}.{Episode Title}.{Quality Title}.{PREFERRED WORDS}-{Release Group}`
Both the EXTENDED and AMZN would be after Quality.

New Supported Renaming Pattern:
`{Series Title}.s{season:00}e{episode:00}.{Episode Title}.{Preferred Words:Editions}.{Quality Title}.{Preferred Words:SourceService}-{Release Group}`

`{Preferred Words}` is still available and includes all defined preferred words, so we're not losing functionality, just adding.

markus101 commented on src/NzbDrone.Core/Profiles/Releases/PreferredWordService.cs:
```diff
@@ -82,5 +84,41 @@ public List<string> GetMatchingPreferredWords(Series series, string title)
         }
-    }
+
+        public Dictionary<string, List<string>> GetMatchingPreferredWordsGroupByProfile(Series series, string title)
```
There should be a single function that returns all preferred words, both ungrouped and grouped, no point in making multiple DB calls.

Alanoll replied:
I removed the original method, and refactored the callers to use the new Dictionary return.
I used the empty string ("") as the key for the "All Matched Words" bucket, since a `{Preferred Words:}` isn't valid and won't be matched.

markus101 reviewed (changes requested):
This definitely needs tests to verify it's functioning properly, the lack of tests for the existing preferred words is an oversight.

markus101 commented on src/NzbDrone.Core/Profiles/Releases/PreferredWordService.cs:
```diff
@@ -76,11 +79,29 @@ public List<string> GetMatchingPreferredWords(Series series, string title)
+                    if (!string.IsNullOrEmpty(releaseProfile.Name))
+                    {
+                        var profileName = releaseProfile.Name.Trim();
+                        if (!profileWords.ContainsKey(profileName))
```
New line before if.

markus101 commented on src/NzbDrone.Core/Profiles/Releases/PreferredWordService.cs:
```diff
@@ -76,11 +79,29 @@ public List<string> GetMatchingPreferredWords(Series series, string title)
+                    {
+                        profileWords.Add(string.Empty, new List<KeyValuePair<string, int>>());
+                    }
+                    profileWords[string.Empty].AddRange(matchingPairs); // Add the "everything grouping"
```
New line after if block.
Don't use a dictionary for the everything bucket, return a new class that has to properties, one that includes everything and one that includes the grouped values.

Alanoll replied:
Replaced the Dictionary result with a new PreferredWordMatchResults object. The name feels just a tiny bit off, so if you'd get a better name, I'm all for it.

markus101 commented on src/NzbDrone.Core/Profiles/Releases/PreferredWordService.cs:
```diff
@@ -3,14 +3,15 @@
         int Calculate(Series series, string title, int indexerId);
-        List<string> GetMatchingPreferredWords(Series series, string title);
-    }
+        Dictionary<string, List<string>> GetMatchingPreferredWordsGroupByProfile(Series series, string title);
```
The original name is still accurate, it returns both grouped and ungrouped values.

Alanoll replied:
Reverted to the original name, and returned the new result class.

markus101 commented on src/NzbDrone.Core/Profiles/Releases/PreferredWordService.cs:
```diff
@@ -76,11 +79,29 @@ public List<string> GetMatchingPreferredWords(Series series, string title)
+
+                if (matchingPairs.Count > 0)
+                {
+                    if (!string.IsNullOrEmpty(releaseProfile.Name))
```
`releaseProfile.Name.IsNotNullOrEmpty()`

Alanoll replied:
Done.

markus101 commented on src/NzbDrone.Core/Organizer/FileNameBuilder.cs:
```diff
@@ -651,14 +651,26 @@ private void AddIdTokens(Dictionary<string, Func<TokenMatch, string>> tokenHandl
+            tokenHandlers["{Preferred Words}"] = m => {
+
+                var profileName = "";
+                if (m.CustomFormat != null)
```
New line before if, use braces.

Alanoll replied:
Done

markus101 commented on src/NzbDrone.Core/Organizer/FileNameBuilder.cs:
```diff
@@ -651,14 +651,26 @@ private void AddIdTokens(Dictionary<string, Func<TokenMatch, string>> tokenHandl
         }
 
-        private void AddPreferredWords(Dictionary<string, Func<TokenMatch, string>> tokenHandlers, Series series, EpisodeFile episodeFile, List<string> preferredWords = null)
+        private void AddPreferredWords(Dictionary<string, Func<TokenMatch, string>> tokenHandlers, Series series, EpisodeFile episodeFile, Dictionary<string, List<string>> profilePreferredWords = null)
```
`profilePreferredWords` -> `preferredWords`

Alanoll replied:
I reverted the name change up the hierarchy up through the interface.

Alanoll commented:
Just checking in to see if there were any other changes requested on this :)

markus101 commented:
Made a small tweak and landed this feature in d4cd4a9549ef41983caf3ff713cb9e3f6970e8a0. Thanks for the work on it!
